service/reconcile-waiter/main.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/red-hat-storage/ocs-client-operator/pkg/utils"
+
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+const intervalSeconds = 5
+
+func main() {
+	// validations
+	operatorNamespace := os.Getenv(utils.OperatorNamespaceEnvVar)
+	if operatorNamespace == "" {
+		klog.Exitf("%s env var is empty", utils.OperatorNamespaceEnvVar)
+	}
+
+	// creation of kube client
+	scheme := runtime.NewScheme()
+	cfg, err := config.GetConfig()
+	if err != nil {
+		klog.Exitf("Failed to get config: %v", err)
+	}
+	cl, err := client.New(cfg, client.Options{Scheme: scheme})
+	if err != nil {
+		klog.Exitf("Failed to create controller runtime client: %v", err)
+	}
+	ctx := context.Background()
+
+	storageClusterCRD := &metav1.PartialObjectMetadata{}
+	storageClusterCRD.SetGroupVersionKind(
+		extv1.SchemeGroupVersion.WithKind("CustomResourceDefinition"),
+	)
+	storageClusterCRD.Name = "storageclusters.ocs.openshift.io"
+
+	// verify presence of StorageCluster CRD
+	if err := cl.Get(ctx, client.ObjectKeyFromObject(storageClusterCRD), storageClusterCRD); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// we did not find StorageCluster CRD and can proceed with reconciliation
+			klog.Info("StorageCluster CRD not found and proceeding with reconciliation of operands")
+			return
+		}
+		klog.Exitf("Failed to verify existence of storagecluster crd: %v", err)
+	} else {
+		// StorageCluster CRD exists, wait till StorageCluster CR is configured in Provider mode
+		storageClusters := &metav1.PartialObjectMetadataList{}
+		storageClusters.SetGroupVersionKind(
+			schema.GroupVersionKind{
+				Group:   "ocs.openshift.io",
+				Version: "v1",
+				Kind:    "StorageCluster",
+			},
+		)
+		// Loop till we are running in Provider mode
+		for {
+			if err := cl.List(ctx, storageClusters, client.InNamespace(operatorNamespace)); err != nil {
+				klog.Exitf("Failed to list storagecluster CRs in namespace %q: %v", operatorNamespace, err)
+			}
+			for _, metadata := range storageClusters.Items {
+				if metadata.GetAnnotations()["mode"] == "provider" {
+					klog.Info("StorageCluster is configured to run in 'Provider' mode and proceeding with reconciliation of operands")
+					return
+				}
+			}
+			if len(storageClusters.Items) == 0 {
+				klog.Info("Waiting for StorageCluster to be deployed")
+			} else {
+				klog.Info("Waiting for StorageCluster deployed mode to be 'Provider'")
+			}
+			time.Sleep(time.Duration(intervalSeconds) * time.Second)
+		}
+	}
```

Comment:
This code should be organized for readability, which means we should minimize the exit flows
My suggestion is (in pseudo-code)
```
1. Read CRD into a var, 
   1.1 sleep then retry read (we should fail on error because it will lead us into a crush loop backoff)
2. If CRD exists 
   2.1 Read storageclsuter CR metadata
   2.2 if not exists or not provider sleep + loop
3. Return peacefully

Reply:
since there is a wait loop (3) has to be in (1.2) and (2.3) which is what I'm doing above.

> will lead us into a crush loop backoff
- yes, I thought of CLBO but left it as kubelet will provide incremental backoff w/ CLBO and so didn't expand the code
- as you also want it to fold into code, I could do that, no qualms

Reply:
>since there is a wait loop (3) has to be in (1.2) and (2.3) which is what I'm doing above.
No, you don't need to there should only be one wait loop and one return

Reply:
done.